### PR TITLE
MM-25887  Ability to disable NPS surveys on a user level

### PR DESCRIPTION
**Pull request by reflog**
#### Summary

implemented per-user disable

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-25887

### CODE REVIEW

**Comment by esethna** [comment]
@hmhealey @jgilliam17 I'd like to target v5.28 for this, can we please test this and post screenshots here

**Comment by esethna** [comment]
Thanks Eli, is there an associated telemetry event for this if a user clicks the link to turn off surveys?

**Comment by reflog** [comment]
@esethna - Yes it includes a new event called 'nps_disable' which will contain a timestamp and userid (Good catch btw, I forgot to commit that piece initially :) )

**Comment by reflog** [comment]
@hmhealey - I'll try to figure something out :)

**Comment by esethna** [comment]
@hmhealey @reflog how do we QA this? Can someone add QA as I don't have permissions on this repo

**Comment by reflog** [comment]
@esethna - it's easily QA'd:
1. install the plugin
2. wait for the survey to arrive
3. click disable
4. validate that the survey doesn't arrive again.

for ease of testing, there's qa-branch that QA previously used to speed up survey arrival.

File: server/api.go [hunk 1]
```diff
@@ -49,6 +54,30 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
 	}
 }
 
+func (p *Plugin) disableForUser(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
```
Comment (larkox) [inline server/api.go:58]: On other plugins we usually check whether this is set, and return a "not authorized" error when this is not set. I see we are not doing this in this plugin at all.
The potential problem is someone not logged in hitting this endpoint, and doing all the operations over an empty string.

File: server/api.go [hunk 2]
```diff
@@ -31,6 +31,11 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
 			Method:  http.MethodPost,
 			Handler: requiresUserId(p.submitScore),
 		},
+		{
+			Path:    "/api/v1/disable_for_user",
```
Comment (larkox) [inline server/api.go:35]: Do we want any way to enable the survey again?

**Review by larkox — APPROVED** [review]
One question, and one improvement that may be out of the scope of this PR.

The rest looks good to me.

File: server/api.go (hunk 2 above)
Comment (reflog) [reply server/api.go:35]: cc @esethna

File: server/api.go (hunk 1 above)
Comment (reflog) [reply server/api.go:58]: good point. i'll add validation

Comment (reflog) [reply server/api.go:58]: @larkox - actually, no need. Requests are wrapped in `requiresUserId` that does the validation

Comment (larkox) [reply server/api.go:58]: I was looking for something like that specifically before making this comment, and I don't know how I didn't see it. :sweat_smile:

File: server/api.go (hunk 2 above)
Comment (esethna) [reply server/api.go:35]: No, probably not necessary. They can still always type to surveybot to give feedback even if automated surveys are disabled

**Review by hmhealey — APPROVED** [review]

**Comment by esethna** [comment]
@lindalumitchell @ogi-m can we get this tested and merged so it can be shipped with v5.28? 

@hmhealey how do we get these changes packaged with v5.28 or is it automated?

**Review by ogi-m — APPROVED** [review]
We'll need to merge first and update the 'qa-build' branch for QA to be able to test.

**Comment by esethna** [comment]
Cool, we can merge then @reflog

**Comment by reflog** [comment]
@esethna @ogi-m done!

**Pull request merged**